pow/ethash.go

```diff
@@ -0,0 +1,612 @@
+type dataset struct {
+	epoch uint64 // Epoch for which this cache is relevant
+
+	dump *os.File  // File descriptor of the memory mapped cache
+	mmap mmap.MMap // Memory map itself to unmap before releasing
+
+	dataset []uint32   // The actual cache data content
+	used    time.Time  // Timestamp of the last use for smarter eviction
+	once    sync.Once  // Ensures the cache is generated only once
+	lock    sync.Mutex // Ensures thread safety for updating the usage time
+}
+
+// generate ensures that the dataset content is generated before use.
+func (d *dataset) generate(dir string, limit int, test bool) {
+	d.once.Do(func() {
+		// If we have a testing dataset, generate and return
+		if test {
+			cache := make([]uint32, 1024/4)
+			generateCache(cache, d.epoch, seedHash(d.epoch*epochLength+1))
+
+			d.dataset = make([]uint32, 32*1024/4)
+			generateDataset(d.dataset, d.epoch, cache)
+
+			return
+		}
+		// If we don't store anything on disk, generate and return
+		csize := cacheSize(d.epoch*epochLength + 1)
+		dsize := datasetSize(d.epoch*epochLength + 1)
+		seed := seedHash(d.epoch*epochLength + 1)
+
+		if dir == "" {
+			cache := make([]uint32, csize/4)
+			generateCache(cache, d.epoch, seed)
+
+			d.dataset = make([]uint32, dsize/4)
+			generateDataset(d.dataset, d.epoch, cache)
+		}
+		// Disk storage is needed, this will get fancy
+		endian := "le"
+		if !isLittleEndian() {
+			endian = "be"
+		}
+		path := filepath.Join(dir, fmt.Sprintf("full-R%d-%x.%s", algorithmRevision, seed, endian))
+		logger := log.New("epoch", d.epoch)
+
+		// Try to load the file from disk and memory map it
+		var err error
+		d.dump, d.mmap, d.dataset, err = memoryMap(path, false)
+		if err == nil {
+			logger.Debug("Loaded old ethash dataset from disk")
+			return
+		}
+		logger.Debug("Failed to load old ethash dataset", "err", err)
+
+		// No previous dataset available, create a new dataset file to fill
+		cache := make([]uint32, csize/4)
+		generateCache(cache, d.epoch, seed)
+
+		if err = os.MkdirAll(filepath.Dir(path), os.ModePerm); err != nil {
+			logger.Error("Failed to create ethash dataset dir", "err", err)
+		} else {
+			// Create a huge temporary empty file to fill with the dataset
+			temp := path + "." + strconv.Itoa(rand.Int())
+
+			dump, err := os.Create(temp)
+			if err != nil {
+				logger.Error("Failed to create temp ethash dataset", "err", err)
+			} else if err = dump.Truncate(int64(dsize)); err != nil {
+				logger.Error("Failed to allocate temp ethash dataset", "err", err)
+			} else {
+				// Dump allocates, map for writing and fill
+				dump.Close()
+
+				dump, mem, dataset, err := memoryMap(temp, true)
+				if err != nil {
+					logger.Error("Failed to map temp ethash dataset", "err", err)
+				}
+				generateDataset(dataset, d.epoch, cache)
+
+				// Generation succeeded, close and move in to final position
+				mem.Unmap()
+				dump.Close()
+
+				os.Rename(temp, path)
+
+				d.dump, d.mmap, d.dataset, err = memoryMap(path, false)
```

Comment:
Please move the memory mapping logic into its own function, e.g.

```go
generateMappedFile(dir, name, size, func (dataset []uint32) { generateDataset(dataset, d.epoch, cache) })
```

This way it can be reused for both cache and DAG.